File: ruby/yahooweather.rb
#----------
#Yahoo weather app using YQL/JSON
#----------
require 'open-uri'
require 'json'

answer
say 'Welcome to the Ruby Yahoo weather reader'

options = { :choices     => "[5 DIGITS]",
            :repeat      => 3,
            :timeout     => 7,
            :onBadChoice => lambda { say 'Invalid entry, please try again.' },
            :onTimeout   => lambda { say 'Timeout, please try again.' },
            :onChoice    => lambda { |choice|

              #Set the URI and our YQL select statement, then encode as a URI
              yahoo_url = 'http://query.yahooapis.com/v1/public/yql?format=json&q='
              query = "SELECT * FROM weather.forecast WHERE location = " + choice.value
              url = URI.encode(yahoo_url + query)
              
              #Fetch the JSON from the YQL API and convert the resulting 
              #JSON data to a Ruby hash
              weather_data = JSON.parse(open(url).read)
              
              #Get the relevant weather channel details and throw them into a hash
              weather_results = weather_data["query"]["results"]["channel"]

              #Speak back the results
              say weather_results["description"]
              say "The wind chill is #{weather_results["wind"]["chill"]} degrees, " + 
                  "the wind speed is #{weather_results["wind"]["speed"]}" 
              say "The forecast is #{weather_results["item"]["forecast"][0]["text"]}, " +
                  "with a high of #{weather_results["item"]["forecast"][0]["high"]} degrees, " +
                  "and a low of #{weather_results["item"]["forecast"][0]["low"]} degrees."
                }
          }

ask 'Enter the ZIP code for a weather check', options
say 'Thats all. Goodbye.'
hangup

File: python/yahooweather.py
import urllib2
from xml.dom import minidom, Node


answer()

say( "Welcome to the python Yahoo weather reader" )
result = ask( "Enter the ZIP code for a weather check", { 'choices' : "[5 DIGITS]" })
if result.name == 'choice' :
    log( "zipCode <" + result.value + ">" )
    urlRead = urllib2.urlopen('http://weather.yahooapis.com/forecastrss?p=' + result.value + '&u=f')
    if urlRead :
        xml = minidom.parse( urlRead )
        if xml :
            for node in xml.documentElement.childNodes :
                if node.nodeName == "channel" :
                    for item_node in node.childNodes :
                        if item_node.nodeName == "description" :
                            description = ""
                            for text_node in item_node.childNodes:
                                if text_node.nodeType == node.TEXT_NODE :
                                    description += text_node.nodeValue
                                if len( description ) > 0:
                                    say( description )
                        if item_node.nodeName == "item" :
                            item = ""
                            for weatherItem_node in item_node.childNodes:
                                if weatherItem_node.nodeName == "title" :
                                    weatherTitle = ""
                                    for weatherText_node in weatherItem_node.childNodes :
                                        weatherTitle += weatherText_node.nodeValue
                                    if len( weatherTitle ) > 0 :
                                        say( weatherTitle )
                                if weatherItem_node.nodeName == "yweather:condition" :
                                    weatherCondition = weatherItem_node.getAttribute( 'temp' )
                                    if len( weatherCondition ) > 0 :
                                        say( "Temperature: " + weatherCondition + " degrees Fahrenheit" )
                                    weatherCondition = weatherItem_node.getAttribute( 'text' )
                                    if len( weatherCondition ) > 0 :
                                        say( weatherCondition )
    else :
        log( "Error getting XML " )
        say( "I am sorry, Error occured while fetching weather." )

say( "Thats all. Goodbye!" )

hangup()

